### ISSUE TITLE
[taichi 0.7.26] Vector does not accumulate properly inside loops

### ISSUE THREAD

**Issue by AmosChenZixuan** [taichi-dev/taichi#2518]
**Describe the bug**
Updating to this newer version of taichi, my previous code stopped working; when using a local accumulating variable to store the sum of some values, it only stores the first value.
P.S. the code is running on cuda.

**To Reproduce**

```py
# inside some kernel
@ti.func
def calc_CM(x):
    # center of mass
    cm = ti.Vector([0.0, 0.0])
    total_mass = 0.0
    for i in x:
        cm += x[i] * M[i]
        total_mass += M[i]
    print(cm, x[0], total_mass)
    cm /= total_mass
    return cm
```
It prints that cm is the same as x[0], instead of the sum of all x. However, total_mass is accumulated correctly.
If I move the print statement inside the loop, then everything is working again.
```py
# inside some kernel
@ti.func
def calc_CM(x):
    # center of mass
    cm = ti.Vector([0.0, 0.0])
    total_mass = 0.0
    for i in x:
        cm += x[i] * M[i]
        total_mass += M[i]
        print(cm, x[0], total_mass)
    cm /= total_mass
    return cm
```

**Log/Screenshots**
print statement is outside the loop 
```
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-ujlv54ih
[Taichi] version 0.7.26, llvm 10.0.0, commit e37bdb5e, linux, python 3.8.8
[Taichi] Starting on arch=cuda
[0.000000, 0.600000] [0.000000, 0.600000] 20.000000
```
print statement is inside the loop (20 lines of the same content)
```
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-ujlv54ih
[Taichi] version 0.7.26, llvm 10.0.0, commit e37bdb5e, linux, python 3.8.8
[Taichi] Starting on arch=cuda
[0.000000, 9.999999] [0.000000, 0.600000] 20.000000
[0.000000, 9.999999] [0.000000, 0.600000] 20.000000
[0.000000, 9.999999] [0.000000, 0.600000] 20.000000
...
```

**Comment by k-ye** [comment]
Suspicious PR: https://github.com/taichi-dev/taichi/pull/2487. @AmesingFlank @BillXu2000 would you help take a look, thanks!

**Comment by AmesingFlank** [comment]
Hi! 

Thank you for submitting this issue.

This is indeed likely a bug in taichi. However,I wasn't able to reproduce this behavior on my end. Could you please include your entire python program instead of just the ti.func ? @AmosChenZixuan

**Comment by AmesingFlank** [comment]
Update:

I was able to reproduce this problem using [this code](https://github.com/AmosChenZixuan/Taichi-PBD/blob/376d90faff2fbe34f542744d186779ce87a20fe5/RigidBody/ShapeMatchingTaichi2.py).

**Comment by bobcao3** [comment]
Fixed by #2519
